### DISCUSSION TITLE
How do I get my internal (or external) dependencies to show up in built package's package.json when using Vite?

### DISCUSSION THREAD

**Discussion by ersinakinci** [nrwl/nx#15517] [General]
Hi! I'm new to nx. What an amazing tool 🤩, thank you Nrwl.io team!

I set up an integrated monorepo with nx, and I created two projects--let's call them `@corp/a` and `@corp/b`. Both use Vite for the build task and Vitest for test.

`@corp/b` depends on `@corp/a` in the project graph, and indeed, I'm importing from `@corp/a`'s code into `@corp/b`'s code.

When I run `nx run-many --target=build --all`, I expect that `@corp/b`'s package.json in `dist` would have `@corp/a` listed as a dependency. However, this doesn't seem to be the case.

It appears that maybe some other plugins, like Rollup and `@nrwl/js`, support this use case using the `updateBuildableProjectDepsInPackageJson` option, but I'm not certain because the documentation is ambiguous. I checked both the docs and the source code for the Vite plugin and there's no equivalent option.

Can someone clarify for me how to get my projects to show up in each other's generated `package.json` files?

**Comment by ersinakinci** [comment]
My current workaround is to manually specify the dependencies in each project's individual `package.json`, which is a little bit of a bummer given all the rich project graph data available to nx.

**Comment by JosefBredereck** [comment]
> It appears that maybe some other plugins, like Rollup and `@nrwl/js`, support this use case using the `updateBuildableProjectDepsInPackageJson` option, but I'm not certain because the documentation is ambiguous. I checked both the docs and the source code for the Vite plugin and there's no equivalent option.

This is also not implemented in completion. There are some issues referencing other packages, since it includes all dependencies down the tree as its own. Means `@corp/a` has all the dependencies of `@corp/b` and `@corp/c`. If the tree is like this:  `@corp/a` ← `@corp/b` ← `@corp/c`.

**Comment by tnrich** [comment]
Seeing the same issue as well.

**Comment by magland** [comment]
I'm facing the same issue. I came across this, and I am going to give it a shot.
https://dev.to/this-is-learning/manage-nx-library-dependencies-with-the-nxdependency-checks-eslint-rule-2lem
https://nx.dev/nx-api/eslint-plugin/documents/dependency-checks

"Nx 16.4 introduces the @nx/dependency-checks ESLint rule to the @nx/linter package for verifying, adding, removing, and updating peerDependencies in the package.json source configuration file of buildable or publishable Nx library projects."

EDIT:

I tried this and it was somewhat tedius to set up manually - and you have to do it separately for each library. I'm confused because the docs say "Library generators from @nx packages will configure this rule automatically when you opt-in for bundler/build setup" -- and that didn't happen for me.

But here's the real problem. Even though the dependencies get added to package.json, vite still bundles all the dependencies together in the dist output anyway! So you need to manually add stuff to vite.config.ts -> build -> rollupOptions -> external. And that needs to be done manually, and is *not* checked by the linter.

So at this point it's not a usable solution, at least for vite.

**Comment by Yuripetusko** [comment]
Hey @ersinakinci have you figured a good pattern for yourself here? I really like vite vs other nx build tools, and working on a monorepo with a lot of publishable libraries that often import each other. Right now they just end up in the bundle, but would love to have them as dependencies or peer-dependencies in dist instead. 

Also I cannot figure out how to make vite's built library to have more extended package.json in the output in dist folder, to include things like `typings` and extend `files` entries in package.json, as I need to include global type declarations there. Would be great to see an example somewhere of someone using nx vite for publishable libraries

**Reply by Yuripetusko** [reply]
Looks like according to this https://github.com/nrwl/nx/issues/13880#issuecomment-1701303212 we just need to manually list all packages in package.json of the library

**Reply by airtonix** [reply]
that comment is really odd. 

if the eslint plugin knows that the package.json in a project is wrong (missing dependencies), why can't there just be an nx plugin that just makes it correct  ?
